python/ray/air/execution/_internal/barrier.py

```diff
@@ -50,7 +65,42 @@ def error(self, *data):
         Args:
             *data: Error data to be cached. Can be obtained via :meth:`get_errors`.
         """
-        raise NotImplementedError
+        self._errors.append(data)
+        self._check_first_error()
+        self._check_completion()
+
+    def _check_first_error(self):
+        if self._has_error:
+            # Already fired callback
+            return
+
+        num_errors = self.num_errors
+        if num_errors:
+            # First error arrived
+            self._has_error = True
+
+            # Invoke callback
+            if self._on_first_error:
+                self._on_first_error(self)
+
+    def _check_completion(self):
+        if self._completed:
+            # Already fired completion callback
+            return
+
+        num_results = self.num_results
+        num_errors = self.num_errors
+
+        if (
+            # max_results is set and number of received results exceeds it
+            self._max_results is not None
```

Comment:
actually, a barrier usually always fires. it may be cleaner if folks can't peek into the internals of a barrier.
like for example, keep the API extremely simple, so we only have
```
results = barrier.wait([list_of_futures])
```
what's the use case of having an infinite sized barrier, or the first error callback?